File: server/app/routers/display.py
import json
from pathlib import Path
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from server.app.config import Config
from server.app.database import get_db
from server.app.models import Layout, Cache, IcsCalendar
from server.app.schemas import DisplayResponse, DisplayWidgetResponse

router = APIRouter(tags=["display"])

_config: Optional[Config] = None


def set_config(config: Config) -> None:
    global _config
    _config = config


def _get_refresh_interval() -> int:
    if _config is None:
        return 60
    settings_path = _config.db_path.parent / "settings.json"
    if settings_path.exists():
        try:
            settings = json.loads(settings_path.read_text())
            return settings.get("display_refresh_interval", 60)
        except (json.JSONDecodeError, OSError):
            pass
    return 60


def _get_cache_key(widget) -> str | None:
    """Determine cache key based on widget type and config.

    For calendar widgets with calendar_sources, returns None (handled by
    _get_calendar_cache_keys instead).
    """
    wtype = widget.widget_type
    config = widget.config or {}
    if wtype == "weather":
        lat = config.get("lat")
        lon = config.get("lon")
        if lat is not None and lon is not None:
            return f"weather_{lat}_{lon}"
    elif wtype == "calendar":
        if config.get("calendar_sources"):
            return None  # Multi-source handled separately
        calendar_ids = sorted(config.get("calendar_ids", ["primary"]))
        days_ahead = config.get("days_ahead", 7)
        return f"google_calendar_{'_'.join(calendar_ids)}_{days_ahead}"
    elif wtype == "photos":
        album_id = config.get("album_id")
        if album_id:
            return f"google_photos_album_{album_id}"
    return None


def _get_calendar_cache_keys(config: dict) -> list[tuple[str, str]]:
    """For calendar widgets with calendar_sources, return list of (source_label, cache_key) tuples."""
    calendar_sources = config.get("calendar_sources", [])
    days_ahead = config.get("days_ahead", 7)
    keys = []
    google_ids = []
    for cs in calendar_sources:
        if cs["type"] == "google":
            google_ids.append(cs["id"])
        elif cs["type"] == "ics":
            keys.append((f"ics:{cs['id']}", f"ics_calendar_{cs['id']}"))
    if google_ids:
        sorted_ids = sorted(google_ids)
        cache_key = f"google_calendar_{'_'.join(sorted_ids)}_{days_ahead}"
        for gid in sorted_ids:
            keys.append((f"google:{gid}", cache_key))
    return keys


def _merge_calendar_data(config: dict, cache_entries: dict) -> dict | None:
    """Merge events from multiple calendar sources, apply colors, sort by start."""
    calendar_sources = config.get("calendar_sources", [])
    if not calendar_sources:
        return None

    colors = config.get("colors", {})
    days_ahead = config.get("days_ahead", 7)
    all_events = []

    # Collect Google calendar events
    google_ids = [cs["id"] for cs in calendar_sources if cs["type"] == "google"]
    if google_ids:
        sorted_ids = sorted(google_ids)
        cache_key = f"google_calendar_{'_'.join(sorted_ids)}_{days_ahead}"
        cached = cache_entries.get(cache_key)
        if cached and "events" in cached:
            for event in cached["events"]:
                # Apply color for each Google source
                # Google events don't have a source label, apply first matching color
                for gid in google_ids:
                    color = colors.get(f"google:{gid}")
                    if color:
                        event = {**event, "color": color}
                        break
                all_events.append(event)

    # Collect ICS calendar events
    for cs in calendar_sources:
        if cs["type"] == "ics":
            cache_key = f"ics_calendar_{cs['id']}"
            cached = cache_entries.get(cache_key)
            if cached and "events" in cached:
                color = colors.get(f"ics:{cs['id']}")
                for event in cached["events"]:
                    if color:
                        event = {**event, "color": color}
                    all_events.append(event)

    if not all_events:
        return None

    # Sort by start time
    all_events.sort(key=lambda e: e.get("start", ""))
    return {"events": all_events}


@router.get("/api/display", response_model=DisplayResponse)
def get_display(db: Session = Depends(get_db)):
    layout = db.query(Layout).filter(Layout.is_active == True).first()
    if not layout:
        raise HTTPException(status_code=404, detail="No active layout")

    # Query all ICS calendars for auto-inclusion in unconfigured calendar widgets
    all_ics = db.query(IcsCalendar).all()

    # Compute needed cache keys from the active layout's widgets, then query only those
    cache_keys = {}
    multi_source_configs = {}  # widget_id -> effective config with resolved sources
    all_needed_keys = set()

    for widget in layout.widgets:
        config = widget.config or {}
        if widget.widget_type == "calendar":
            effective_config = dict(config)
            # Auto-include all ICS calendars when no explicit sources configured
            if not effective_config.get("calendar_sources") and not effective_config.get("calendar_ids") and all_ics:
                effective_config["calendar_sources"] = [{"type": "ics", "id": ic.id} for ic in all_ics]
            if effective_config.get("calendar_sources"):
                multi_source_configs[widget.id] = effective_config
                cal_keys = _get_calendar_cache_keys(effective_config)
                for _label, cache_key in cal_keys:
                    all_needed_keys.add(cache_key)
            else:
                key = _get_cache_key(widget)
                if key:
                    cache_keys[widget.id] = key
                    all_needed_keys.add(key)
        else:
            key = _get_cache_key(widget)
            if key:
                cache_keys[widget.id] = key
                all_needed_keys.add(key)

    if all_needed_keys:
        cache_entries = {c.source: c.data for c in db.query(Cache).filter(Cache.source.in_(all_needed_keys)).all()}
    else:
        cache_entries = {}

    widgets = []
    for widget in layout.widgets:
        if widget.id in multi_source_configs:
            data = _merge_calendar_data(multi_source_configs[widget.id], cache_entries)
        else:
            cache_key = cache_keys.get(widget.id)
            data = cache_entries.get(cache_key) if cache_key else None
        widgets.append(DisplayWidgetResponse(
            id=widget.id,
            widget_type=widget.widget_type,
            config=widget.config,
            data=data,
            position_x=widget.position_x,
            position_y=widget.position_y,
            width=widget.width,
            height=widget.height,
        ))

    return DisplayResponse(
        layout={
            "id": layout.id,
            "name": layout.name,
            "columns": layout.columns,
            "row_height": layout.row_height,
            "theme": layout.theme,
        },
        widgets=widgets,
        refresh_interval=_get_refresh_interval(),
    )

File: src/server/routes/display.ts
import { readFileSync, existsSync } from 'fs'
import { dirname, join } from 'path'
import type { FastifyInstance } from 'fastify'
import { getCacheMultiple } from '../db/queries/cache.js'
import { listIcsCalendars } from '../db/queries/ics-calendars.js'
import { listEnabledScheduleRules } from '../db/queries/schedule-rules.js'
import { evaluateSchedule } from '../services/schedule.js'
import { loadOrCreateKey } from '../services/encryption.js'
import { getValidAccessToken } from '../services/google-auth.js'
import { extractAlbumToken } from '../services/apple-photos.js'
import type { DisplayResponse, DisplayWidgetResponse } from '@shared/types.js'

interface WidgetRow {
  id: number
  layout_id: number
  widget_type: string
  config: string
  position_x: number
  position_y: number
  width: number
  height: number
  created_at: string
  updated_at: string
}

interface LayoutRow {
  id: number
  name: string
  columns: number
  row_height: number
  is_active: number
  theme: string
  created_at: string
  updated_at: string
}

interface CalendarSource {
  type: 'google' | 'ics'
  id: string | number
}

function loadSettings(config: { dbPath: string }): Record<string, unknown> {
  const path = join(dirname(config.dbPath), 'settings.json')
  if (existsSync(path)) {
    try {
      return JSON.parse(readFileSync(path, 'utf-8'))
    } catch {
      return {}
    }
  }
  return {}
}

function getRefreshInterval(config: { dbPath: string }): number {
  const settings = loadSettings(config)
  return (settings.display_refresh_interval as number) ?? 60
}

/**
 * Determine cache key based on widget type and config.
 * For calendar widgets with calendar_sources, returns null (handled by getCalendarCacheKeys).
 */
export function getCacheKey(widget: { widget_type: string; config: Record<string, unknown> }): string | null {
  const { widget_type, config } = widget
  if (widget_type === 'weather') {
    const lat = config.lat
    const lon = config.lon
    if (lat != null && lon != null) {
      return `weather_${lat}_${lon}`
    }
  } else if (widget_type === 'calendar') {
    if (config.calendar_sources) {
      return null // Multi-source handled separately
    }
    const calendarIds = (config.calendar_ids as string[] | undefined) ?? ['primary']
    const sortedIds = [...calendarIds].sort()
    const daysAhead = (config.days_ahead as number) ?? 7
    return `google_calendar_${sortedIds.join('_')}_${daysAhead}`
  } else if (widget_type === 'photos') {
    const photosSource = config.photos_source as string | undefined
    const pickerSessionId = config.picker_session_id
    const icloudAlbumUrl = config.icloud_album_url as string | undefined

    if (photosSource === 'apple' && icloudAlbumUrl) {
      const token = extractAlbumToken(icloudAlbumUrl)
      if (token) {
        return `apple_photos_${token}`
      }
    } else if (pickerSessionId) {
      // Google Photos (explicit source='google' or legacy without photos_source)
      return `google_photos_picker_${pickerSessionId}`
    }
  }
  return null
}

/**
 * For calendar widgets with calendar_sources, return list of [sourceLabel, cacheKey] tuples.
 */
export function getCalendarCacheKeys(config: Record<string, unknown>): [string, string][] {
  const calendarSources = config.calendar_sources as CalendarSource[] | undefined
  if (!calendarSources) return []

  const daysAhead = (config.days_ahead as number) ?? 7
  const keys: [string, string][] = []
  const googleIds: string[] = []

  for (const cs of calendarSources) {
    if (cs.type === 'google') {
      googleIds.push(String(cs.id))
    } else if (cs.type === 'ics') {
      keys.push([`ics:${cs.id}`, `ics_calendar_${cs.id}`])
    }
  }

  if (googleIds.length > 0) {
    const sortedIds = [...googleIds].sort()
    const cacheKey = `google_calendar_${sortedIds.join('_')}_${daysAhead}`
    for (const gid of sortedIds) {
      keys.push([`google:${gid}`, cacheKey])
    }
  }

  return keys
}

/**
 * Merge events from multiple calendar sources, apply colors, sort by start.
 */
export function mergeCalendarData(
  config: Record<string, unknown>,
  cacheEntries: Map<string, unknown>,
): Record<string, unknown> | null {
  const calendarSources = config.calendar_sources as CalendarSource[] | undefined
  if (!calendarSources || calendarSources.length === 0) return null

  const colors = (config.colors as Record<string, string>) ?? {}
  const daysAhead = (config.days_ahead as number) ?? 7
  const allEvents: Record<string, unknown>[] = []

  // Collect Google calendar events
  const googleIds = calendarSources
    .filter((cs) => cs.type === 'google')
    .map((cs) => String(cs.id))

  if (googleIds.length > 0) {
    const sortedIds = [...googleIds].sort()
    const cacheKey = `google_calendar_${sortedIds.join('_')}_${daysAhead}`
    const cached = cacheEntries.get(cacheKey) as { events?: Record<string, unknown>[] } | undefined
    if (cached?.events) {
      for (const event of cached.events) {
        let e = event
        for (const gid of googleIds) {
          const color = colors[`google:${gid}`]
          if (color) {
            e = { ...e, color }
            break
          }
        }
        allEvents.push(e)
      }
    }
  }

  // Collect ICS calendar events
  for (const cs of calendarSources) {
    if (cs.type === 'ics') {
      const cacheKey = `ics_calendar_${cs.id}`
      const cached = cacheEntries.get(cacheKey) as { events?: Record<string, unknown>[] } | undefined
      if (cached?.events) {
        const color = colors[`ics:${cs.id}`]
        for (const event of cached.events) {
          if (color) {
            allEvents.push({ ...event, color })
          } else {
            allEvents.push(event)
          }
        }
      }
    }
  }

  if (allEvents.length === 0) return null

  // Sort by start time
  allEvents.sort((a, b) => {
    const aStart = (a.start as string) ?? ''
    const bStart = (b.start as string) ?? ''
    return aStart.localeCompare(bStart)
  })

  return { events: allEvents }
}

export async function displayRoutes(app: FastifyInstance): Promise<void> {
  const config = (app as unknown as { config: { dbPath: string; secretKeyPath: string } }).config
  const db = (app as unknown as { db: import('better-sqlite3').Database }).db

  app.get('/api/display', async (_request, reply) => {
    // Check schedule
    const settings = loadSettings(config)
    const schedulingEnabled = settings.scheduling_enabled === true

    let layout: LayoutRow | undefined

    if (schedulingEnabled) {
      const rules = listEnabledScheduleRules(db)
      const scheduleResult = evaluateSchedule(rules, new Date())

      if (scheduleResult) {
        if (scheduleResult.display_power === 'off') {
          // Display off - return minimal response
          const response: DisplayResponse = {
            layout: null,
            widgets: [],
            refresh_interval: getRefreshInterval(config),
            display_power: 'off',
          }
          return response
        }
        // Schedule says show a specific layout
        layout = db.prepare('SELECT * FROM layouts WHERE id = ?').get(scheduleResult.layout_id) as LayoutRow | undefined
      }
    }

    // Fallback: use manually-activated layout
    if (!layout) {
      layout = db.prepare('SELECT * FROM layouts WHERE is_active = 1').get() as LayoutRow | undefined
    }

    if (!layout) {
      reply.code(404).send({ error: 'No active layout' })
      return
    }

    // Query all ICS calendars for auto-inclusion
    const allIcs = listIcsCalendars(db)

    // Get widgets for active layout
    const widgetRows = db.prepare('SELECT * FROM widgets WHERE layout_id = ?').all(layout.id) as WidgetRow[]

    // Compute needed cache keys
    const cacheKeys = new Map<number, string>() // widget_id -> cache_key
    const multiSourceConfigs = new Map<number, Record<string, unknown>>() // widget_id -> effective config
    const allNeededKeys = new Set<string>()

    for (const row of widgetRows) {
      const widgetConfig = JSON.parse(row.config) as Record<string, unknown>
      if (row.widget_type === 'calendar') {
        const effectiveConfig = { ...widgetConfig }
        // Auto-include all ICS calendars when no explicit sources configured
        if (!effectiveConfig.calendar_sources && !effectiveConfig.calendar_ids && allIcs.length > 0) {
          effectiveConfig.calendar_sources = allIcs.map((ic) => ({ type: 'ics', id: ic.id }))
        }
        if (effectiveConfig.calendar_sources) {
          multiSourceConfigs.set(row.id, effectiveConfig)
          const calKeys = getCalendarCacheKeys(effectiveConfig)
          for (const [, cacheKey] of calKeys) {
            allNeededKeys.add(cacheKey)
          }
        } else {
          const key = getCacheKey({ widget_type: row.widget_type, config: widgetConfig })
          if (key) {
            cacheKeys.set(row.id, key)
            allNeededKeys.add(key)
          }
        }
      } else {
        const key = getCacheKey({ widget_type: row.widget_type, config: widgetConfig })
        if (key) {
          cacheKeys.set(row.id, key)
          allNeededKeys.add(key)
        }
      }
    }

    // Batch-query cache
    const cacheEntries = allNeededKeys.size > 0
      ? getCacheMultiple(db, [...allNeededKeys])
      : new Map<string, unknown>()

    // Build widget responses
    const widgets: DisplayWidgetResponse[] = widgetRows.map((row) => {
      const widgetConfig = JSON.parse(row.config) as Record<string, unknown>
      let data: Record<string, unknown> | null = null

      if (multiSourceConfigs.has(row.id)) {
        data = mergeCalendarData(multiSourceConfigs.get(row.id)!, cacheEntries)
      } else {
        const cacheKey = cacheKeys.get(row.id)
        if (cacheKey) {
          data = (cacheEntries.get(cacheKey) as Record<string, unknown>) ?? null
        }
      }

      return {
        id: row.id,
        widget_type: row.widget_type as DisplayWidgetResponse['widget_type'],
        config: widgetConfig,
        data,
        position_x: row.position_x,
        position_y: row.position_y,
        width: row.width,
        height: row.height,
      }
    })

    const response: DisplayResponse = {
      layout: {
        id: layout.id,
        name: layout.name,
        columns: layout.columns,
        row_height: layout.row_height,
        theme: JSON.parse(layout.theme),
      },
      widgets,
      refresh_interval: getRefreshInterval(config),
      display_power: 'on',
    }

    return response
  })

  app.get<{ Querystring: { url: string } }>('/api/photos/proxy', async (request, reply) => {
    const { url } = request.query
    if (!url || !url.startsWith('https://lh3.googleusercontent.com/')) {
      reply.code(400).send({ error: 'Invalid photo URL' })
      return
    }

    const settings = loadSettings(config)
    const clientId = (settings.google_client_id as string) ?? ''
    const clientSecret = (settings.google_client_secret as string) ?? ''

    let key: Buffer
    try {
      key = loadOrCreateKey(config.secretKeyPath)
    } catch {
      reply.code(502).send({ error: 'Encryption key unavailable' })
      return
    }

    const accessToken = await getValidAccessToken(db, key, clientId, clientSecret)
    if (!accessToken) {
      reply.code(502).send({ error: 'Google not connected' })
      return
    }

    const photoUrl = `${url}=w1920-h1080`

    try {
      const resp = await fetch(photoUrl, {
        headers: { Authorization: `Bearer ${accessToken}` },
        redirect: 'follow',
        signal: AbortSignal.timeout(15000),
      })

      if (!resp.ok) {
        request.log.warn(`Photo proxy failed: ${resp.status}`)
        reply.code(resp.status).send({ error: 'Photo fetch failed' })
        return
      }

      const contentType = resp.headers.get('content-type') ?? 'image/jpeg'
      const buffer = Buffer.from(await resp.arrayBuffer())

      reply
        .header('content-type', contentType)
        .header('cache-control', 'private, max-age=3000')
        .send(buffer)
    } catch (err) {
      request.log.warn(`Photo proxy request error: ${err}`)
      reply.code(502).send({ error: 'Photo fetch failed' })
    }
  })
}
